### PR TITLE
feat: add Iraq parser

### PR DESCRIPTION
**Pull request by byronwilliams**
Adds the Iraq parser

#2850 #2852

### CODE REVIEW

**Comment by alixunderplatz** [comment]
Hi @byronwilliams , congratulations on your first contribution in this repo 👍  Looking forward to having Iraq back on the map soon! :)

In case you hadn't seen it, we used to have data for Iraq in the past, until the data source went offline a whole while ago, see old PR #1447 and the old parser code here:
https://github.com/tmrowco/electricitymap-contrib/pull/1447/files

The imports/exports with the neigbouring regions were also fetched in the past. You can adjust the fetch_exchange part of the old parser to bring them back to the map as well, the ids of the exchanges still seem to be the same 😃 

grafik
grafik

"fetch_consumption" for total system load was also used back then

BTW: I just changed the URL in the old parser #1447 to the new one and got all data with the old code 😄
{'zoneKey': 'IQ', 'datetime': datetime.datetime(2020, 12, 26, 0, 52, 18, tzinfo=tzfile('Asia/Baghdad')), 'production': {'gas': 7128.7, 'hydro': 514.2, 'oil': 2625.7}, 'storage': {}, 'source': 'Iraqi Power System'}
{'zoneKey': 'IQ', 'datetime': datetime.datetime(2020, 12, 26, 0, 52, 18, tzinfo=tzfile('Asia/Baghdad')), 'consumption': 11521.3, 'source': 'Iraqi Power System'}
{'sortedZoneKeys': 'IQ->IR', 'datetime': datetime.datetime(2020, 12, 26, 0, 52, 18, tzinfo=tzfile('Asia/Baghdad')), 'netFlow': -1242, 'source': 'Iraqi Power System'}
{'sortedZoneKeys': 'IQ->IQ-KUR', 'datetime': datetime.datetime(2020, 12, 26, 0, 52, 18, tzinfo=tzfile('Asia/Baghdad')), 'netFlow': 0, 'source': 'Iraqi Power System'}
But nevermind, if you could implement the exchange data to your code, it'll be perfect 👍

**Comment by corradio** [comment]
Hello both! Thanks for the doing this! Do we prefer reusing the old code or should we proceed with this PR? Any pros or cons?

**Comment by alixunderplatz** [comment]
@corradio I guess the old parser with the updated URL is the fastest option to bring it back to the map including fetch_exchange & fetch_consumption. new parser code seems shorter. it's up to you 😄

**Comment by corradio** [comment]
> @corradio I guess the old parser with the updated URL is the fastest option to bring it back to the map including fetch_exchange & fetch_consumption. new parser code seems shorter. it's up to you 😄

Hi @byronwilliams!
Are you able to help us bring back up the exchanges as well?
I suggest taking a look at the PR that @alixunderplatz referred to.
Thank you for the help.

Olivier

**Comment by corradio** [comment]
friendly bump @byronwilliams

**Comment by byronwilliams** [comment]
@corradio - because you asked nicely :smile:  - I repurposed the code from the previous PR as you suggested

**Comment by corradio** [comment]
> @corradio - because you asked nicely 😄 - I repurposed the code from the previous PR as you suggested

Awesome! Thank you so much. @alixunderplatz can you confirm we can merge this?

File: parsers/IQ.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    result = template_response(zone_key, timestamp, DATA_SOURCE)
+
+    for production_type, ids in CELL_MAPPING.items():
+        for id in ids:
+            result["production"][production_type] += data["d_" + id]
+
+    return [result]
+
+
+def fetch_exchange(zone_key1, zone_key2, session=None, target_datetime=None, logger=None):
```
Comment (skovhus) [inline parsers/IQ.py:95]: If this should be used, then we need to update the exchange.json file with a reference to this parser for each of the exchanges.

File: parsers/IQ.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        "source": source,
+    }
+
+def fetch_data(r):
+    resp = r.get(LIVE_PRODUCTION_API_URL)
+    data = resp.json()
+
+    timestamp = arrow.get(data["lastmodified"], "HH:mm:ss A DD-MM-YYYY", tzinfo=tz.gettz('Asia/Baghdad'))
+
+    return data["d"], timestamp
```
Comment (skovhus) [inline parsers/IQ.py:43]: As `datetime` must be returned as native `datetime.datetime` object

```suggestion
    return data["d"], timestamp.datetime
```

**Review by skovhus — COMMENTED** [review]
Thanks for fixing this. 

I'm curious how you mapped the various ids (e.g. `d_1226`) to something meaningful.

**Comment by skovhus** [comment]
After fixing the `datetime` issue and updating the zones.json file I get the following output from `test_parser.py`:

Screenshot 2021-02-25 at 14 54 05

I haven't tested the exchanges.

**Comment by corradio** [comment]
Updated in https://github.com/tmrowco/electricitymap-contrib/pull/3035 instead. Thanks @byronwilliams for doing all the work!

**Pull request closed without merging**